Feature/create-ata-map-widget

phillipet opened the pull request:
A custom Flutter Google Map has some features:
1. Specify current location of device
2. Pin any marker on map and move it
3. Show circle around marked position with deviation radius
4. Calulate distance between current location and marked position

Example code :

#admin user
```dart
Widget build(BuildContext context) {
    return MaterialApp(
        title: 'Flutter Google Maps Demo',
        home: Column(
          children: <Widget>[
            Expanded(
              child: ATAMap(
                isMoveableMarker: true,
                authRange: 100,
              ),
            ),
            Text('XXX'),
            Text('YYY'),
          ],
        ));
  }
```
#normal user
```dart
ATAMap(
      isMoveableMarker: false,
      titleMarkedPosition: 'Office Location',
      markedLat: 10.7440878,
      markedLng: 106.7007886,
)
```

phillipet committed:
add ATAMap widget

phillipet committed:
update document for ata_map widget

phillipet committed:
show lat,lng of marked position for admin user

phillipet committed:
fix condition showing title marker is lat and lng

phillipet committed:
update doc & clean code

phillipet committed:
update doc, focus to main features

phillipet committed:
move DEVIATION_RADIUS to constructor

don't use const in widget, only adjust this value from constructor(able received from db...)

phillipet committed:
clean code

hemvanh commented on lib/ui/widgets/ata_map.dart:
```diff
@@ -0,0 +1,186 @@
+  final bool isMoveableMarker;
+  final String titleMarkedPosition;
+
+  ATAMapState({this.currentMarkedLat, this.currentMarkedLng, this.centerMapLat, this.centerMapLng, this.isMoveableMarker, this.deviationRadius, this.titleMarkedPosition}) {
```
no need to add params to State constructor just to pass values from Widget to its state.

inside the State<> class, we can always access the widget's props through widget.markedLat

hemvanh commented on lib/ui/widgets/ata_map.dart:
```diff
@@ -0,0 +1,185 @@
+          markerId: MarkerId(point.toString()),
+          position: point,
+          infoWindow: InfoWindow(
+              title: isMoveableMarker == true ? '$currentMarkedLat, $currentMarkedLng' : '$titleMarkedPosition',
```
isMoveableMarker? '$currentMarkedLat, $currentMarkedLng' : '$titleMarkedPosition',
why need to check isMoveableMarker == true, isMoveableMarker is already a bool ?

hemvanh commented on lib/ui/widgets/ata_map.dart:
```diff
@@ -0,0 +1,185 @@
+          icon: markedLocationIcon,
+          draggable: isMoveableMarker,
+          onDragEnd: ((newPoint) {
+            if (isMoveableMarker == true) {
```
just if(isMoveableMarker )  -> enough

hemvanh reviewed (changes requested):
revise your code

phillipet committed:
clean code condition with bool variable as javascript(lasted dart sdk)

phillipet committed:
remove constructor of StatefulWidget, use widget key word to access prop

-fix loading maker and get current location on map -created event
-move setState function correctly
-prevent long press when map is getting current location

phillipet committed:
clean code

hemvanh commented on lib/ui/widgets/ata_map.dart:
```diff
@@ -0,0 +1,181 @@
+      }
+    }
+    setState(() {
+      if (_markers.length == 0) _addMarker(LatLng(currentMarkedLat, currentMarkedLng));
```
_addMarker already have setState() inside, so no need to setState() before calling it

hemvanh commented on lib/ui/widgets/ata_map.dart:
```diff
@@ -0,0 +1,181 @@
+          onDragEnd: ((newPoint) {
+            if (widget.isMoveableMarker) {
+              _addMarker(newPoint);
+              currentMarkedLat = newPoint.latitude;
```
currentMarkedLat = newPoint.latitude;  -> no need to set again, coz inside _addMarker() already has set it at the beginning

hemvanh reviewed (changes requested):
revise your code

phillipet committed:
clean code
